### PR TITLE
[semantic-arc-opts] Implement @owned phi web elimination for phi webs with a single phi node that only have copy_value introducers.

### PR DESCRIPTION
**Pull request by gottesmm**
This is the most simple initial version that I can commit. The hope is that this will help to bring this up in a nice way.

I am going to handle the multiple phi node and load [copy] case later to reduce
code churn.

<rdar://problem/56720436>

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci benchmark

**Comment by gottesmm** [comment]
@swift-ci test source compatibility

**Comment by gottesmm** [comment]
This is also one of the most simple cases, so I am not necessarily expecting a large change in perf. Will take what I can get though = p.

**Comment by gottesmm** [comment]
I think the release failure is a problem on the bot. Saw this:

```
16:02:11 <unknown>:0: note: Cannot read '/Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/build/compat_macos/swift-macosx-x86_64/./lib/swift/iphonesimulator/layouts-x86_64.yaml'
16:02:11 Stack dump:
```

**Comment by gottesmm** [comment]
Reproduced the failure locally. I wonder if it is non-deterministic.

**Comment by gottesmm** [comment]
Debug source compat passed. I think the release one hit a race. Going to rerun.

**Comment by gottesmm** [comment]
@swift-ci test source compatibility

**Comment by gottesmm** [comment]
One of the failures in release suite looks real and actually related to the logic. Other has to do with my not being aggressive enough using autogenerated locations.

**Comment by gottesmm** [comment]
Looking at the failures now.

**Comment by gottesmm** [comment]
Found the core store issue. I used a propagated a branch location onto a begin_borrow. I just changed it to use an autogenerated location.

**Comment by gottesmm** [comment]
Both turned out to be the loc issue.

**Comment by gottesmm** [comment]
@swift-ci test source compatibility

**Comment by gottesmm** [comment]
@swift-ci test source compatibility

**Comment by gottesmm** [comment]
@swift-ci test source compatibility

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@atrick was just thinking. I think the same approach here would work with tuples and other aggregate forming instructions. They act as a phi in a certain sense of their operand's ownership.

**Comment by gottesmm** [comment]
I talked with Andy offline. I am going to refactor parts of it according to his asks in a forthcoming commit.

**Review by atrick — COMMENTED** [review]
Just a few insignificant comments... This and all the offline feedback can go into a separate PR

File: lib/SILOptimizer/Transforms/SemanticARCOpts.cpp [hunk 1]
```diff
@@ -627,6 +789,240 @@ static llvm::cl::opt<bool>
...
+
+  {
+    // If we made a change, set that we assume we are at fixed point and then
+    // re-run the worklist so that we can
+    // properly seeded the ARC peephole map.
+    assumingAtFixedPoint = true;
+    SWIFT_DEFER { assumingAtFixedPoint = false; };
+
+    // Add everything in visitedSinceLastMutation to the worklist.
+    for (auto opt : visitedSinceLastMutation) {
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SILOptimizer/Transforms/SemanticARCOpts.cpp]: I'm also going to include the full type to make it clear it is Optional<SILValue>. I think it makes the code easier to understand.

[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SILOptimizer/Transforms/SemanticARCOpts.cpp]: #30320

File: lib/SILOptimizer/Transforms/SemanticARCOpts.cpp [hunk 2]
```diff
@@ -83,10 +93,32 @@ class LiveRange {
...
+    return unknownConsumingUses.front();
+  }
+
+  OwnedValueIntroducer getIntroducer() const { return introducer; }
+
   ArrayRef<Operand *> getNonConsumingForwardingUses() const {
     return generalForwardingUses;
   }
 
+  void convertOwnedGeneralForwardingUsesToGuaranteed();
```
[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SILOptimizer/Transforms/SemanticARCOpts.cpp]: I have a patch but it conflicts my other fix patch. I am going to wait for that to land then get to this.

**Comment by davezarzycki** [comment]
Hi @gottesmm – This breaks my Fedora 31 box (x86_64). Do you have a quick fix or can we revert this?
```
FAIL: Swift(linux-x86_64) :: SILOptimizer/semantic-arc-opts.sil (724 of 12842)
******************** TEST 'Swift(linux-x86_64) :: SILOptimizer/semantic-arc-opts.sil' FAILED ********************
Script:
--
: 'RUN: at line 1';   /home/dave/b/u/t/bin/sil-opt -target x86_64-unknown-linux-gnu  -module-cache-path '/home/dave/b/u/t/swift-test-results/x86_64-unknown-linux-gnu/clang-module-cache'  -module-name Swift -enable-sil-verify-all -semantic-arc-opts /home/dave/s/u/swift/test/SILOptimizer/semantic-arc-opts.sil | '/usr/bin/python2.7' '/home/dave/s/u/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/home/dave/b/u/t/tools/swift' --sanitize SOURCE_DIR='/home/dave/s/u/swift' --use-filecheck '/home/dave/b/u/t/bin/FileCheck'  /home/dave/s/u/swift/test/SILOptimizer/semantic-arc-opts.sil
--
Exit Code: 2

Command Output (stderr):
--
SIL verification failed: instruction isn't dominated by its operand: properlyDominates(valueI, I)
Verifying instruction:
     %7 = begin_borrow %0 : $Klass                // users: %3, %8
->   %3 = begin_borrow %7 : $Klass                // user: %4
     br bb5(%3 : $Klass)                          // id: %4
In function:
// copy_guaranteed_three_copy_simple
sil [ossa] @copy_guaranteed_three_copy_simple : $@convention(thin) (@guaranteed Klass) -> () {
// %0                                             // user: %7
bb0(%0 : @guaranteed $Klass):
  cond_br undef, bb1, bb2                         // id: %1

bb1:                                              // Preds: bb0
  cond_br undef, bb3, bb4                         // id: %2

bb2:                                              // Preds: bb0
  %3 = begin_borrow %7 : $Klass                   // user: %4
  br bb5(%3 : $Klass)                             // id: %4

bb3:                                              // Preds: bb1
  %5 = begin_borrow %5 : $Klass                   // users: %6, %5
  br bb5(%5 : $Klass)                             // id: %6

bb4:                                              // Preds: bb1
  %7 = begin_borrow %0 : $Klass                   // users: %3, %8
  br bb5(%7 : $Klass)                             // id: %8

// %9                                             // user: %10
bb5(%9 : @guaranteed $Klass):                     // Preds: bb4 bb3 bb2
  end_borrow %9 : $Klass                          // id: %10
  %11 = tuple ()                                  // user: %12
  return %11 : $()                                // id: %12
} // end sil function 'copy_guaranteed_three_copy_simple'

Stack dump:
0.      Program arguments: /home/dave/b/u/t/bin/sil-opt -target x86_64-unknown-linux-gnu -module-cache-path /home/dave/b/u/t/swift-test-results/x86_64-unknown-linux-gnu/clang-module-cache -module-name Swift -enable-sil-verify-all -semantic-arc-opts /home/dave/s/u/swift/test/SILOptimizer/semantic-arc-opts.sil 
1.      While evaluating request ExecuteSILPipelineRequest(Run pipelines { Pass List Pipeline } on SIL for Swift.Swift)
2.      While running pass #84 SILFunctionTransform "SemanticARCOpts" on SILFunction "@copy_guaranteed_three_copy_simple".
 for /home/dave/s/u/swift/test/SILOptimizer/semantic-arc-opts.sil:1961:13
3.      While verifying SIL function "@copy_guaranteed_three_copy_simple".
 for /home/dave/s/u/swift/test/SILOptimizer/semantic-arc-opts.sil:1961:13
 #0 0x00000000067f8424 llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/home/dave/b/u/t/bin/sil-opt+0x67f8424)
 #1 0x00000000067f6777 llvm::sys::RunSignalHandlers() (/home/dave/b/u/t/bin/sil-opt+0x67f6777)
 #2 0x00000000067f885a SignalHandler(int) (/home/dave/b/u/t/bin/sil-opt+0x67f885a)
 #3 0x00007ffff7fa2b20 __restore_rt (/lib64/libpthread.so.0+0x14b20)
 #4 0x00007ffff6152625 raise (/lib64/libc.so.6+0x3c625)
 #5 0x00007ffff613b8d9 abort (/lib64/libc.so.6+0x258d9)
 #6 0x00000000030f9ec8 (/home/dave/b/u/t/bin/sil-opt+0x30f9ec8)
 #7 0x0000000003112520 (anonymous namespace)::SILVerifier::visitSILInstruction(swift::SILInstruction*) (/home/dave/b/u/t/bin/sil-opt+0x3112520)
 #8 0x00000000030fb7fc (anonymous namespace)::SILVerifier::visitSILBasicBlock(swift::SILBasicBlock*) (/home/dave/b/u/t/bin/sil-opt+0x30fb7fc)
 #9 0x00000000030f5a37 swift::SILFunction::verify(bool) const (/home/dave/b/u/t/bin/sil-opt+0x30f5a37)
#10 0x0000000002769335 swift::SILPassManager::runPassOnFunction(unsigned int, swift::SILFunction*) (/home/dave/b/u/t/bin/sil-opt+0x2769335)
#11 0x0000000002769fdc swift::SILPassManager::runFunctionPasses(unsigned int, unsigned int) (/home/dave/b/u/t/bin/sil-opt+0x2769fdc)
#12 0x0000000002767408 swift::SILPassManager::executePassPipelinePlan(swift::SILPassPipelinePlan const&) (/home/dave/b/u/t/bin/sil-opt+0x2767408)
#13 0x0000000002767398 swift::ExecuteSILPipelineRequest::evaluate(swift::Evaluator&, swift::SILPipelineExecutionDescriptor) const (/home/dave/b/u/t/bin/sil-opt+0x2767398)
#14 0x00000000027647b3 swift::SimpleRequest<swift::ExecuteSILPipelineRequest, bool (swift::SILPipelineExecutionDescriptor), (swift::CacheKind)0>::evaluateRequest(swift::ExecuteSILPipelineRequest const&, swift::Evaluator&) (/home/dave/b/u/t/bin/sil-opt+0x27647b3)
#15 0x0000000002771af8 llvm::Expected<swift::ExecuteSILPipelineRequest::OutputType> swift::Evaluator::getResultUncached<swift::ExecuteSILPipelineRequest>(swift::ExecuteSILPipelineRequest const&) (/home/dave/b/u/t/bin/sil-opt+0x2771af8)
#16 0x0000000002767614 swift::executePassPipelinePlan(swift::SILModule*, swift::SILPassPipelinePlan const&, bool, swift::irgen::IRGenModule*) (/home/dave/b/u/t/bin/sil-opt+0x2767614)
#17 0x000000000240eaaf main (/home/dave/b/u/t/bin/sil-opt+0x240eaaf)
#18 0x00007ffff613d1a3 __libc_start_main (/lib64/libc.so.6+0x271a3)
#19 0x0000000002406c2e _start (/home/dave/b/u/t/bin/sil-opt+0x2406c2e)
FileCheck error: '-' is empty.
FileCheck command line:  /home/dave/b/u/t/bin/FileCheck /home/dave/s/u/swift/test/SILOptimizer/semantic-arc-opts.sil

--

********************

Testing Time: 53.15s
********************
Failing Tests (1):
    Swift(linux-x86_64) :: SILOptimizer/semantic-arc-opts.sil

  Expected Passes    : 11101
  Expected Failures  : 24
  Unsupported Tests  : 1716
  Unexpected Failures: 1


```

**Pull request merged**